### ISSUE TITLE
Missing nil checking for the array element type

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#13602]
The compiler should not allow default-initialized arrays whose element type is non-nilable.

```chpl
class MyClass {var x:int;}
var global: [1..2] owned MyClass; // should be an error
```

**Comment by mppf** [comment]
Another example spun off from #8597:

``` chapel
class Turtle {
  var x: int;
  var D: domain(string) = {"down"};
  var A: [D] owned Turtle;
}

var worldTurtle = new Turtle();
writeln(worldTurtle);
```

Please not that resizing an array of non-nilable types will not work (e.g. `D += "test"` in the above) because the new elements can't be default-initialized.

**Comment by dlongnecke-cray** [comment]
I've wrestled with this issue while working on #14367. My proposal in that thread https://github.com/chapel-lang/chapel/issues/14367#issuecomment-580542261 is that for now, we should prohibit arrays of nil-able classes, and then work to expand those constraints as much as possible before the release.

**Comment by bradcray** [comment]
It seems as though, in the case of rectangular arrays, we could support arrays of nilable classes as long as the array is initialized (since there's no way to partially initialize a rectangular array, as there is an associative array).

**Comment by vasslitvinov** [comment]
I will try Brad's proposal above.

I read it to support arrays of NON-nilable classes in those special cases. Nilable classes do not present problems with resizing, I think.

**Comment by dlongnecke-cray** [comment]
Agh! I meant to say _non-nillable_, sorry for the confusion.

**Comment by vasslitvinov** [comment]
Our rectangular domain maps have these fields:
```chpl
  var targetLocDom: domain(rank);
  var targetLocales: [targetLocDom] locale;
```

This is unsound because `targetLocDom` can be resized and the newly-added indices in `targetLocales` will need to be default-initialized.

One way to handle it is to make targetLocales's elements nilable. I don't like that because we'd have to use a lot of `!` checks. 

Another way is to make `targetLocDom` a `const`. This will not work, unfortunately, because we have ex. `Block.dsiAssign()` that assigns `targetLocDom`. If it is a `const`, `dsiAssign` will not compile.

This leaves me with the solution of addign an exception. I suggest following @dlongnecke-cray's lead, who has already considered (and implemented?) an exception for associative arrays of locales.

Thoughts?

**Comment by bradcray** [comment]
I'd lean toward making it const and avoiding reliance on dsiAssign.  My guess is that we use it today for copy initializers and that it's rarely, if ever, used by the assignment overload for real codes.

Do locales not have a default value, though?  I thought at some point we'd discussed making Locales[0] the default value.

**Comment by dlongnecke-cray** [comment]
I had to touch `modules/dists/ReplicatedDist.chpl` already while working on #14367, and it looks like I ended up just making the associative array `targetLocales` nilable in the end...

```chapel
const targetLocales: [targetLocDom] locale?;
```

Would you end up having to touch all of `modules/dists` with this change?

I ended up choosing not to make an exception for the `locale` type because it didn't really accomplish anything for me. It just so happens that `_locale` is a class, but the real problem is default initializing a non-nilable class...

**Comment by bradcray** [comment]
Oh, I think this is how the conversation went last time:  I suggested that I thought locales should be changed to a record type that wraps the class type in order to support default values and the like.  But then we never did it...

**Comment by vasslitvinov** [comment]
It so happens that we do not use `=(_distribution,_distribution)` anywhere in our tests. So doing away with dsiAssign on distribution classes should be an option. Which should resolve the issue with `var targetLocales` fields.

So - how do we go about removing `=(_distribution,_distribution)` ? I don't think we can "deprecate" it because then the above solution will not work, i.e. we still have to support dsiAssign. Given that probably nobody heard about it, how about I just yank it? Or switch it to `compilerError()` ?

**Comment by bradcray** [comment]
I'm OK with that, though I do think that the better solution (albeit more expensive to implement) would be to change the locale type to a record-wrapped-class type and give it a default value of Locales[0].

**Comment by mppf** [comment]
> Do locales not have a default value, though? I thought at some point we'd discussed making Locales[0] the default value.

Maybe we should spin this off into a different issue? In particular I am wondering why we wouldn't make `here` be the default value.

**Comment by bradcray** [comment]
> In particular I am wondering why we wouldn't make here be the default value.

I'm open to that as well.

**Comment by vasslitvinov** [comment]
BTW we should expand the scope of this check, as well as the one in #15094, to all non-default-initializable types. Ex. ones that do not pass `isDefaultInitializable()` in Types.chpl.

**Comment by vasslitvinov** [comment]
Resolved by #15044 and #15224.
